packages/components/src/components/Calculator/index.tsx

```diff
@@ -0,0 +1,153 @@
+import React, { FC, ReactNode, useEffect, useReducer, useState } from "react";
+
+import { SqCalculator, SqValuePath } from "@quri/squiggle-lang";
+import { Env } from "@quri/squiggle-lang";
+
+import { PlaygroundSettings } from "../PlaygroundSettings.js";
+import { SqValueWithContext } from "../../lib/utility.js";
+
+import {
+  processAllFieldCodes,
+  updateAndProcessFieldCode,
+} from "./asyncActions.js";
+
+import { Action, useViewerContext } from "../SquiggleViewer/ViewerProvider.js";
+import {
+  CalculatorAction,
+  CalculatorState,
+  calculatorHash,
+  calculatorReducer,
+  hasSameCalculator,
+  initialCalculatorState,
+} from "./calculatorReducer.js";
+import { CalculatorUI } from "./calculatorUI.js";
+
+type Props = {
+  value: SqCalculator;
+  environment: Env;
+  settings: PlaygroundSettings;
+  valueWithContext: SqValueWithContext;
+  renderValue: (
+    value: SqValueWithContext,
+    settings: PlaygroundSettings
+  ) => ReactNode;
+};
+
+//We backup the calculator state in the viewer context, for when the user navigates away and back
+const adjustedReducer =
+  (path: SqValuePath, viewerContextDispatch: (action: Action) => void) =>
+  (state: CalculatorState, action: CalculatorAction) => {
+    const newState = calculatorReducer(state, action);
+    viewerContextDispatch({
+      type: "CALCULATOR_UPDATE",
+      payload: {
+        path: path,
+        calculator: newState,
+      },
+    });
+    return newState;
+  };
+
+export const Calculator: FC<Props> = ({
```

Comment:
I think this file is messier than it could be. That said, I think cleaning it up could be a fair bit of work. Maybe a good fit for a few hooks or something.